runtime/interpreter/value.go

```diff
@@ -6487,6 +6544,72 @@ func (v *CompositeValue) ConformsToDynamicType(
 	return true
 }
 
+func (v *CompositeValue) Fields() *StringValueOrderedMap {
+	v.ensureFieldsLoaded()
+	return v.fields
+}
+
+func (v *CompositeValue) Location() common.Location {
+	v.ensureMetaInfoLoaded()
+	return v.location
+}
+
+func (v *CompositeValue) QualifiedIdentifier() string {
+	v.ensureMetaInfoLoaded()
+	return v.qualifiedIdentifier
+}
+
+func (v *CompositeValue) Kind() common.CompositeKind {
+	v.ensureMetaInfoLoaded()
+	return v.kind
+}
+
+// ensureMetaInfoLoaded ensures loading the meta information of this composite value.
+// If the meta info is already loaded, then calling this function won't have any effect.
+// Otherwise, the values are decoded form the cached raw-content.
+//
+// Meta info includes:
+//    - location
+//    - qualifiedIdentifier
+//    - kind
+//
+func (v *CompositeValue) ensureMetaInfoLoaded() {
+	if v.content == nil {
+		return
+	}
+
+	err := decodeCompositeMetaInfo(v, v.content)
+	if err != nil {
+		panic(err)
+	}
+
+	// Raw content is no longer needed. Clear the cache and free-up the memory.
+	v.content = nil
+}
+
+// ensureFieldsLoaded ensures loading the fields of this composite value.
+// If the fields are already loaded, then calling this function won't have any effect.
+// Otherwise, the fields are decoded form the cached raw-fields-content.
+//
+func (v *CompositeValue) ensureFieldsLoaded() {
+	// First ensure the fields content is extracted out.
+	v.ensureMetaInfoLoaded()
+
+	if v.fieldsContent == nil {
+		return
+	}
+
+	err := decodeCompositeFields(v, v.fieldsContent)
+	if err != nil {
+		panic(err)
+	}
+
+	// Path and the fields-content are no longer needed.
```

Comment:
What do you think about postponing the release of v.fieldsContent until it is modified?  That way, we don't have to re-encode it if it's stored without changes.